### PR TITLE
Checkout tooltip

### PR DESCRIPTION
**Pull request by danimoh**
Tooltip with a price breakdown for checkouts.

Would `shopFee` be a better name for `merchantFee` as we also have `shopLogoUrl`?

### CODE REVIEW

File: src/components/PaymentInfoLine.vue [hunk 1]
```diff
@@ -63,14 +142,20 @@ function fiatAmountInfoValidator(value: any) {
...
+    private static readonly REFERENCE_RATE_UPDATE_INTERVAL = 60000; // every minute
+    private static readonly RATE_DEVIATION_THRESHOLD = .05;
 
     @Prop({type: Object, required: true, validator: cryptoAmountInfoValidator}) public cryptoAmount!: CryptoAmountInfo;
     @Prop({type: Object, validator: fiatAmountInfoValidator}) public fiatAmount?: FiatAmountInfo;
+    // Note that merchantFee and networkFee have no effect if fiatAmount is not set, as the tooltip in which they appear
+    // is only visible when fiatAmount is set. As the fiatAmount was only introduced in the v2 checkout request in the
+    // hub, the tooltip and merchantFee and networkFee are thus never visible for v1 checkout requests. This is ok
+    // though as the merchantFee also only exists in v2 and the free-service-info doesn't make too much sense for
+    // nimiq.shop or the nimiq voting app which are currently the main apps using the v1 checkout.
```
Comment (sisou) [inline src/components/PaymentInfoLine.vue:156]: I would leave out the last sentence of the comment here. The tooltip is not displayed for non-fiat checkouts, period. Talking about other apps will not age well, and `This is ok though` is also an opinion, not a fact, so not very helpful I think.

**H**ub should be capitalized.

File: src/components/PaymentInfoLine.vue [hunk 2]
```diff
@@ -63,14 +142,20 @@ function fiatAmountInfoValidator(value: any) {
...
+    @Prop({validator: amountValidator}) public merchantFee?: number | bigint | BigInteger;
```
Comment (sisou) [inline src/components/PaymentInfoLine.vue:157]: I see you take the merchant fee as an absolute Luna value. The setting in the WooCommerce plugin is in percent, though. And it's also displayed in percent, so shouldn't the prop be simply in percent?

File: src/components/PaymentInfoLine.vue [hunk 3]
```diff
@@ -138,6 +307,66 @@ export default PaymentInfoLine;
...
+        border: unset;
+        border-top: 1px solid currentColor;
+        opacity: .2;
+    }
+
+    .price-tooltip .total {
+        /* The total row is on purpose not part of the grid as the label is shorter and the value longer */
+        display: flex;
+        justify-content: space-between;
+    }
```
Comment (sisou) [inline src/components/PaymentInfoLine.vue:357]: In the designs, the total crypto amount is `font-weight: bold`.

File: src/components/PaymentInfoLine.vue [hunk 4]
```diff
@@ -82,12 +167,95 @@ class PaymentInfoLine extends Vue {
...
+    private get effectiveRate() {
+        if (!this.fiatAmount) return null;
+        // Note: precision loss should be acceptable here
+        return this.fiatAmount.amount / (Number(this.cryptoAmount.amount) / (10 ** this.cryptoAmount.decimals));
+    }
+
+    private get formattedMerchantFee() {
+        if (this.merchantFee === null || this.merchantFee === undefined) return null;
+        // Note: precision loss should be acceptable here
+        const merchantFeePercent = Number(this.merchantFee) / Number(this.cryptoAmount.amount) * 100;
```
Comment (sisou) [inline src/components/PaymentInfoLine.vue:201]: This calculation is not correct.

Example: `cryptoAmount = 2000, merchantFee = 500`. This calculation computes a 25% merchant rate.

But actually, the merchantFee is already _included_ in the cryptoAmount, so you have to subtract it first. This means the actual item price is `2000 - 500 = 1500`. So the actual merchant rate would be `500 / 1500 = 33%`.

Note that this calculation is not needed anymore if the merchant rate prop is accepted in percent.

File: src/components/PaymentInfoLine.vue [hunk 5]
```diff
@@ -82,12 +167,95 @@ class PaymentInfoLine extends Vue {
...
+        // Round to two decimals. Always ceil to avoid displaying a lower fee than charged.
+        return `${Math.ceil(merchantFeePercent * 1e2) / 1e2}%`;
```
Comment (sisou) [inline src/components/PaymentInfoLine.vue:203]: No need to write 100 in e-notation, it's the same length of three characters, but more confusing :laughing: 
```suggestion
        return `${Math.ceil(merchantFeePercent * 100) / 100}%`;
```

File: src/components/PaymentInfoLine.vue (hunk 1 above)
Comment (danimoh) [reply src/components/PaymentInfoLine.vue:156]: I made minor adjustments to the comment but see the last sentence as essential as it gives the reasoning why this component does not show a tooltip for checkouts without fiat amount.

File: src/components/PaymentInfoLine.vue (hunk 2 above)
Comment (danimoh) [reply src/components/PaymentInfoLine.vue:157]: It's a relative value now.

File: src/components/PaymentInfoLine.vue (hunk 5 above)
Comment (danimoh) [reply src/components/PaymentInfoLine.vue:203]: It was actually written like this to make the intention clearer ;)

**Comment by danimoh** [comment]
Good catch detecting the _higher_/_lower_, _better_/_worse_ confusion.
The updated implementation resembles the current state of discussions.

File: src/components/PaymentInfoLine.vue [hunk 6]
```diff
@@ -63,14 +146,20 @@ function fiatAmountInfoValidator(value: any) {
...
-    }
+    private static readonly REFERENCE_RATE_UPDATE_INTERVAL = 60000; // every minute
+    private static readonly RATE_DEVIATION_THRESHOLD = .1;
 
     @Prop({type: Object, required: true, validator: cryptoAmountInfoValidator}) public cryptoAmount!: CryptoAmountInfo;
     @Prop({type: Object, validator: fiatAmountInfoValidator}) public fiatAmount?: FiatAmountInfo;
+    // Note that vendorMarkup and networkFee have no effect if fiatAmount is not set, as the tooltip in which they
+    // appear is only visible when fiatAmount is set. As the fiatAmount was only introduced in the v2 checkout request
+    // in the Hub, the tooltip and vendorMarkup and networkFee are thus never visible for v1 checkout requests. This
+    // should be ok though as the vendorMarkup also only exists in v2 and the free-service-info doesn't make too much
```
Comment (sisou) [inline src/components/PaymentInfoLine.vue:159]: I still think "This should be ok though" should not be put into a comment in the sourc code.

File: src/components/PaymentInfoLine.vue [hunk 7]
```diff
@@ -1,17 +1,95 @@
...
+                class="price-tooltip"
+            >
+                <template v-slot:trigger>
+                    <FiatAmount :currency="fiatAmount.currency" :amount="fiatAmount.amount" />
+                </template>
+                <template v-slot:default>
+                    <div class="price-breakdown">
+                        <label>Order amount</label>
+                        <FiatAmount :currency="fiatAmount.currency" :amount="fiatAmount.amount" />
+                        <template v-if="vendorMarkup || vendorMarkup === 0">
```
Comment (sisou) [inline src/components/PaymentInfoLine.vue:38]: Was it agreed with the designers that the vendor markup is even displayed when 0%?

File: src/components/PaymentInfoLine.vue [hunk 8]
```diff
@@ -36,24 +36,25 @@
...
-                        :class="{ 'nq-orange': rateDeviation >= constructor.RATE_DEVIATION_THRESHOLD }"
+                    <div v-if="isBadRate || Math.abs(rateDeviation) >= constructor.RATE_DEVIATION_THRESHOLD"
+                        :class="{ 'nq-orange': isBadRate }"
                         class="rate-info info"
                     >
                         You are paying approx.
+                        {{ isBadRate && rateDeviation < 0 ? 'only' : '' }}
                         {{ formattedRateDeviation }}
                         {{ rateDeviation > 0 ? 'more' : 'less' }}
                         than at the current market rate (coingecko.com).
```
Comment (sisou) [inline src/components/PaymentInfoLine.vue]: @Yukiioz This is going to be fun for i18n :smile:

File: src/components/PaymentInfoLine.vue [hunk 9]
```diff
@@ -1,17 +1,95 @@
...
+                    </div>
+                    <div v-if="isBadRate || Math.abs(rateDeviation) >= constructor.RATE_DEVIATION_THRESHOLD"
+                        :class="{ 'nq-orange': isBadRate }"
+                        class="rate-info info"
+                    >
+                        You are paying approx.
+                        {{ isBadRate && rateDeviation < 0 ? 'only' : '' }}
+                        {{ formattedRateDeviation }}
+                        {{ rateDeviation > 0 ? 'more' : 'less' }}
+                        than at the current market rate (coingecko.com).
```
Comment (sisou) [inline src/components/PaymentInfoLine.vue:60]: @Yukiioz This is going to be fun for i18n :smile:

File: src/components/PaymentInfoLine.vue (hunk 7 above)
Comment (danimoh) [reply src/components/PaymentInfoLine.vue:38]: Several implementation decisions were taken without involvement of the designers.
They have unfortunately been very silent when it came to details of the tooltip.
I asked them to give final feedback now but think the choices made are reasonable.

**Comment by danimoh** [comment]
The warning as it appears is intended.
The vendor claims to give a discount of 50% but in fact you're paying **only** 7.6% below market value, therefore a warning is issued.
Surely, the warning could be made more clear for that case, but I think it's good enough.
The warning on inaccurate discounts serves to show merchants that they shouldn't try promising discounts they're not actually granting. In the real world, user should hopefully not see this case.

File: src/components/PaymentInfoLine.vue (hunk 7 above)
Comment (danimoh) [reply src/components/PaymentInfoLine.vue:38]: They did not complain about the handling of the 0% case.

**Comment by danimoh** [comment]
Has been adapted to deisgner feedback and should be final now.

**Pull request merged**